go-controller/pkg/ovn/egressip.go

```diff
@@ -1228,6 +608,14 @@ func (oc *DefaultNetworkController) deleteEgressIPAssignments(name string, statu
 				// we can continue here since this pod was not managed by this statusToRemove
 				continue
 			}
+			// this statusToRemove was managing at least one pod, hence let's tear down the setup for this status
```

Comment:
yes, it was done because of your suggestion actually hehe.
Remember you, me and Pat had a call few weeks ago about a race bug? When we add things we look at the cache to see if it exists or not and add it accordingly, but when we delete things we weren't checking the cache. Hence we were removing setup for stale statuses which removed setup for other newly added statuses. So not totally for `if pod exists in cache` ; more for `does status exist in cache`.

We had so many stale NAT bugs that we clean too well :)
See
```
routerName := util.GetGatewayRouterFromNode(status.Node)
		natPred := func(nat *nbdb.NAT) bool {
			// TODO(tssurya): investigate if we should be deleting NATs only from status.Node
			return nat.ExternalIDs["name"] == name && nat.ExternalIP == status.EgressIP
		}
```
within `deleteEgressIPStatusSetup`
So we had two options, 1) either check cache before triggering delete OR 2) delete NAT based on status.Node - we were hesitant to do the first approach.

Reply:
I am sometimes having a bad event race only in CI - its the same as we discussed here; https://github.com/ovn-org/ovn-kubernetes/pull/3386#discussion_r1196790084
EIP is assigned to node1:
```
2023-05-22T16:54:26.7175003Z I0522 16:54:26.717311   21445 egressip.go:1206] Patching status on EgressIP egressip: [{node1 192.168.126.101}]
```
Label changes to node2:
```
2023-05-22T16:54:26.7208352Z I0522 16:54:26.720620   21445 egressip.go:1206] Patching status on EgressIP egressip: [{node2 192.168.126.101}]
```
Add pod setup for node1 status:
```
2023-05-22T16:54:26.7314476Z I0522 16:54:26.731207   21445 egressip.go:574] Adding pod egress IP status: {node1 192.168.126.101} for EgressIP: egressip and pod: egressip-namespace/egress-pod/[10.128.0.15/23]
```
Add pod setup for node2 status:
```
2023-05-22T16:54:26.7698191Z I0522 16:54:26.769604   21445 egressip.go:574] Adding pod egress IP status: {node2 192.168.126.101} for EgressIP: egressip and pod: egressip-namespace/egress-pod/[10.128.0.15/23]
```
and finally the stupid delete for node1 status happens:
```
2023-05-22T16:54:26.7921426Z I0522 16:54:26.791895   21445 egressip.go:636] Deleting pod egress IP status: {node1 192.168.126.101} for EgressIP: egressip
```
and because of https://github.com/ovn-org/ovn-kubernetes/blob/0e395d5a831930fd9f5e91d3303bf6d636da6136/go-controller/pkg/ovn/egressip.go#L2620 where we don't check for node, we remove all the NATs of all the nodes - this is not a problem in IC world where we only have one node but its failing my UTs for multiple nodes in same zone its happened twice now:  https://github.com/ovn-org/ovn-kubernetes/actions/runs/5048097599/jobs/9055855508?pr=3386 and https://github.com/ovn-org/ovn-kubernetes/actions/runs/5048097599/jobs/9055855508?pr=3386. I am going to re-add the status.Node match to remove NATs that are only on the node we care about!